cmake/modules/RootMacros.cmake

```diff
@@ -393,28 +393,39 @@ function(ROOT_GENERATE_DICTIONARY dictionary)
   endforeach()
 
   #---Build the names for library, pcm and rootmap file ----
+  set(library_target_name)
   if(dictionary MATCHES "^G__")
-    string(SUBSTRING ${dictionary} 3 -1 deduced_arg_module)
+    string(REGEX REPLACE "^G__(.*)" "\\1"  library_target_name ${dictionary})
+    if (ARG_MULTIDICT)
+      string(REGEX REPLACE "(.*)32$" "\\1"  library_target_name ${library_target_name})
+    endif (ARG_MULTIDICT)
   else()
-    get_filename_component(deduced_arg_module ${dictionary} NAME_WE)
+    get_filename_component(library_target_name ${dictionary} NAME_WE)
   endif()
+  if (ARG_MODULE)
+    if (NOT ${ARG_MODULE} STREQUAL ${library_target_name})
+#      message(AUTHOR_WARNING "The MODULE argument ${ARG_MODULE} and the deduced library name "
+#        "${library_target_name} mismatch. Deduction stem: ${dictionary}.")
+      set(library_target_name ${ARG_MODULE})
+    endif()
+  endif(ARG_MODULE)
```

Comment:
This should be the default, and the else can contain the deduction code above. No need to set if the deduction fails or to check that they are the same...

Reply:
Here the important bit was the diagnostics. I can probably get rid of it altogether. The real underlying problem is that `ROOT_GENERATE_DICTIONARY` allows the user to specify the name of the dictionary which will be produced. I think, instead, we should just specify the library which the dictionary describes and the rest should be automatically computed.

Reply:
There is no underlying problem here. The user must be able to specify the name of the dictionary, and it's important because you can have a target with multiple dictionaries and deducing a name in that case is prone to bugs. The dictionary is just a source file that users can add to their library or binary and the name convention is something that ROOT should not impose. You need to think more from the point of view of users of ROOT, and less of ROOT itself. We cannot change macros used externally at will. If we suddenly remove this functionality, users will certainly complain.

Reply:
The deduction is here since 5 years. We deduce the name because it may not be specified as you said. Do you want to change that or you want it to be just more efficient — if we move the deduction in the else part we will not be able to diagnose such mismatch. If we don’t want to diagnose (without imposing naming convention) that’s fine.

We should work very hard to hide the name of the dictionary. It contains intricate information for the library such as locations of header files and rdict files and so on. We have an option how to build multiple dictionaries for a single library—if that does not work we should fix it. Allowing the user to decide which dictionary gets attached to which library and when is a source of big issues and it is very subtle bugs. Essentially the dictionary is a ROOT implementation detail which can affect the core of the framework majorly. 

We should encourage people to use the ROOT_STANDARD_LIBRARY_PPACKAGE and we should not use a standalone ROOT_GENERATE_DICTIONARY standalone.